CodeStarterCamp_Week4/main.swift

```diff
@@ -80,52 +86,166 @@ class BodyCondition {
+    guard bodyCondition.피로도 < 100 else {
+        throw ActivityError.overFatigue
+    }
 })
-let 오래달리기: Activity = Activity(name: "오래달리기", action: { bodyCondition in
-    print("<<\(오래달리기.name)을(를) 시작합니다>>")
-    bodyCondition.상체근력 += Int.random(in: 5...10)
-    bodyCondition.하체근력 += Int.random(in: 5...10)
-    bodyCondition.근지구력 += Int.random(in: 20...30)
-    bodyCondition.피로도 += Int.random(in: 20...30)
+
+let 오래달리기: Activity = Activity(name: "오래달리기", action: { bodyCondition throws in
+    doActivity(bodyCondition: bodyCondition, activityName: 오래달리기.name, min상체근력: 5, max상체근력: 10, min하체근력: 5, max하체근력:  10, min근지구력: 20, max근지구력: 30, min피로도: 20, max피로도: 30)
+    guard bodyCondition.피로도 < 100 else {
+        throw ActivityError.overFatigue
+    }
 })
+
 let 동적휴식: Activity = Activity(name: "동적휴식", action: { bodyCondition in
-    print("<<\(동적휴식.name)을(를) 시작합니다>>")
-    bodyCondition.피로도 -= Int.random(in: 5...10)
-})
-let 동적음주: Activity = Activity(name: "동적음주", action: { bodyCondition in
-    print("<<\(동적음주.name)을(를) 시작합니다>>")
-    bodyCondition.상체근력 -= Int.random(in: 5...10)
-    bodyCondition.하체근력 -= Int.random(in: 20...30)
-    bodyCondition.근지구력 -= Int.random(in: 5...10)
-    bodyCondition.피로도 += Int.random(in: 10...20)
+    doActivity(bodyCondition: bodyCondition, activityName: 동적휴식.name, min피로도: -10, max피로도: -5)
+    guard bodyCondition.피로도 < 100 else {
+        throw ActivityError.overFatigue
+    }
 })
 
+let 동적음주: Activity = Activity(name: "동적음주", action: { bodyCondition throws in
+    doActivity(bodyCondition: bodyCondition, activityName: 동적음주.name, min상체근력: -10, max상체근력: -5, min하체근력: -30, max하체근력: -20, min근지구력: -10, max근지구력: -5, min피로도: 10, max피로도: 20)
+    guard bodyCondition.피로도 < 100 else {
+        throw ActivityError.overFatigue
+    }
+})
 
+// Testing...
 var myBodyCondition: BodyCondition = BodyCondition()
+let hellRoutine = Routine(name: "hellRoutine", activities: [윗몸일으키기, 동적휴식, 스쿼트])
+var countRoutines = 0
 
-// testing for each action...
-윗몸일으키기.action(myBodyCondition)
-오래달리기.action(myBodyCondition)
-동적휴식.action(myBodyCondition)
-동적음주.action(myBodyCondition)
-myBodyCondition.printCurrentCondition()
+print("루틴을 몇 번 반복할까요?")
 
-// testing for 피로도 < 0 ...
-for _ in 1...10 {
-    동적휴식.action(myBodyCondition)
+do {
+    repeat {
+        do {
+            var inputCountRoutines = readLine()
+            countRoutines = try checkInputCount(inputCountRoutines: inputCountRoutines)
+        } catch ActivityError.invalidInput {
+            print("잘못된 입력 형식입니다. 다시 입력해주세요.")
+        }
+    } while countRoutines == 0
+
+    print("--------------")
+    
+    for countRoutine in 1...countRoutines {
+        print("\(changeNumToOrdinal(num: countRoutine))번째 \(hellRoutine.name)을(를) 시작합니다.")
+        for routineActivity in hellRoutine.activities {
+            try routineActivity.action(myBodyCondition)
+        }
+    }
+} catch ActivityError.overFatigue {
+    print("피로도가 100 이상입니다. 루틴을 중단합니다.")
 }
```

Comment:
이 부분에선 기능분리가 필요 해보입니다. 먼저 루틴을 시작하는 부분이 있는데 이 부분은 Routine 타입의 메서드로 만들어도 좋을 것 같아요. 입력 받는 부분도 메서드로 묶어줄 수 있을 것 같아요. 모든 정리가 완료 되면 최종적으로 입력받는 메서드, 루틴을 실행하는 메서드, 현재컨디션 출력메서드 이렇게 세개의 메서드를 호출하면 지금과 같은 출력이 나오도록 수정 해볼까요 ?

Reply:
`takeInputCount`, `startRoutines` 함수로 기능을 분리하였습니다.
다만 오류 처리 과정에서 문제가 있었는데요,
다음 `startRoutines` 함수 내부입니다.
```
do {
         for countRoutine in 1...countRoutines {
             print("\(changeNumToOrdinal(num: countRoutine))번째 \(routine.name)을(를) 시작합니다.")
             for routineActivity in routine.activities {
                 try routineActivity.action(myBodyCondition)
             }
         }
     } catch ActivityError.overFatigue {
         print("피로도가 100 이상입니다. 루틴을 중단합니다.")
     }
```

`try routineActivity.action(myBodyCondition)` 실행문에서는 `ActivityError.overFatigue` 오류만 던질 수 있기 때문에, `ActivityError.overFatigue` 오류에 대한 catch구문만 만들어주면 된다고 생각합니다.

하지만 저렇게 구현할 경우 함수 내에서 모든 오류 케이스를 커버하지 않는다는 에러가 발생합니다.
`에러 문구 : Errors thrown from here are not handled because the enclosing catch is not exhaustive`

따라서 catch 구문을 다음과 같이 수정할 수 밖에 없었는데요,
```
catch {
         print("피로도가 100 이상입니다. 루틴을 중단합니다.")
     }
```
수정 결과 코드는 정상 작동 하지만, 마치 모든 에러에 대해서 `print("피로도가 100 이상입니다. 루틴을 중단합니다.")`를 수행하는것처럼 보여서 찝찝한 느낌입니다... 😢  이런 경우 어떻게 해결해야할지 조언해 주실 수 있을까요? 😄

Reply:
routinActivity.action() 메서드에서 오류를 하나밖에 던지지 않기 때문에 이렇게 처리 해주어도 상관은 없을 것 같습니다.
catch 구문에서는 어떤 에러가 던져질지 정확히 추측할 수 없기 때문에 이런 오류가 발생한 것 같아요. ActivityError가 아닌 다른에러가 발생하면 어떻게 처리 해줄거냐? 이렇게 물어보는 것과 같죠. 우리는 에러가 한가지밖에 존재하지 않는다는 걸 알기 때문에 catch로 간단하게 받아주어도 될 것 같아요. 위처럼 `catch ActivityError.overFatigue` 로 받아주고 싶다면, 또 다른 `catch { }` 블럭을 아래 추가해서 ActivityError.overFatigue 가 아닌 다른 에러가 발생했을 때 실행해줄 구문을 작성해주면 될 것 같아요.(이 구문이 실행될 일은 없다는 걸 알지만요.)